File: glut_opengl_estudos/01_janela_e_triangulo.cpp
/*
Aula 01 — Primeira janela com GLUT + OpenGL (C++)

Objetivo pedagógico:
- Entender o ciclo base de uma aplicação gráfica com GLUT.
- Entender o callback de renderização (`display`).
- Renderizar um triângulo colorido com interpolação de cores.

Compilação (Linux):
  g++ -std=c++17 01_janela_e_triangulo.cpp -o aula01 -lglut -lGL -lGLU
*/

#include <GL/glut.h>
#include "cpp_glut_utils.hpp"

namespace {

void display() {
    // Limpa color buffer (fundo da janela).
    glClear(GL_COLOR_BUFFER_BIT);

    // Modo imediato: cada vértice recebe uma cor.
    // A GPU interpola as cores no interior do triângulo.
    glBegin(GL_TRIANGLES);
    glColor3f(1.0f, 0.0f, 0.0f); glVertex2f(-0.7f, -0.6f);
    glColor3f(0.0f, 1.0f, 0.0f); glVertex2f( 0.7f, -0.6f);
    glColor3f(0.0f, 0.2f, 1.0f); glVertex2f( 0.0f,  0.7f);
    glEnd();

    glFlush();
}

} // namespace

int main(int argc, char** argv) {
    glutInit(&argc, argv);
    aula::configureCloseBehavior();
    glutInitDisplayMode(GLUT_SINGLE | GLUT_RGB);
    glutInitWindowSize(800, 600);
    glutCreateWindow("Aula 01 - Triangulo (C++)");

    glClearColor(0.05f, 0.05f, 0.08f, 1.0f);
    glutDisplayFunc(display);
    glutMainLoop();
    return 0;
}

File: glut_opengl_estudos/01_janela_e_triangulo.py
"""
Aula 01 — Primeira janela com GLUT + OpenGL

Objetivos didáticos:
1) Entender o papel do GLUT (janela, eventos e loop).
2) Entender o callback `display`.
3) Desenhar um triângulo com cores interpoladas.

Como executar:
    python3 01_janela_e_triangulo.py
"""

from OpenGL.GL import (
    GL_COLOR_BUFFER_BIT,
    GL_DEPTH_BUFFER_BIT,
    GL_TRIANGLES,
    glBegin,
    glClear,
    glClearColor,
    glColor3f,
    glEnd,
    glFlush,
    glVertex2f,
)
from OpenGL.GLUT import (
    GLUT_ACTION_GLUTMAINLOOP_RETURNS,
    GLUT_ACTION_ON_WINDOW_CLOSE,
    GLUT_DEPTH,
    GLUT_RGB,
    GLUT_SINGLE,
    glutCreateWindow,
    glutDisplayFunc,
    glutInit,
    glutInitDisplayMode,
    glutInitWindowSize,
    glutMainLoop,
    glutSetOption,
)


def configure_close_behavior() -> None:
    """Tenta evitar `exit()` do processo ao fechar a janela (quando suportado)."""
    try:
        glutSetOption(GLUT_ACTION_ON_WINDOW_CLOSE, GLUT_ACTION_GLUTMAINLOOP_RETURNS)
    except Exception:
        pass


def display() -> None:
    """Callback de desenho.

    O OpenGL imediato (`glBegin`/`glEnd`) é simples para estudo.
    Cada vértice recebe uma cor; a GPU interpola no interior do triângulo.
    """
    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT)

    glBegin(GL_TRIANGLES)
    glColor3f(1.0, 0.0, 0.0)
    glVertex2f(-0.7, -0.6)

    glColor3f(0.0, 1.0, 0.0)
    glVertex2f(0.7, -0.6)

    glColor3f(0.0, 0.2, 1.0)
    glVertex2f(0.0, 0.7)
    glEnd()

    glFlush()


def main() -> None:
    """Inicializa contexto OpenGL e inicia o loop de eventos."""
    glutInit()
    configure_close_behavior()
    glutInitDisplayMode(GLUT_SINGLE | GLUT_RGB | GLUT_DEPTH)
    glutInitWindowSize(800, 600)
    glutCreateWindow(b"Aula 01 - Triangulo")

    glClearColor(0.05, 0.05, 0.08, 1.0)
    glutDisplayFunc(display)

    print("Feche a janela para encerrar a aula 01.")
    glutMainLoop()


if __name__ == "__main__":
    main()
